examples/experimental/dagster-airlift/dagster_airlift/core/serialization/compute.py

```diff
@@ -23,29 +25,73 @@
 from dagster_airlift.core.utils import spec_iterator
 
 
+@record
+class TaskSpecMappingInfo:
+    asset_specs: List[AssetSpec]
+
+    @cached_property
+    def asset_keys(self) -> Set[AssetKey]:
+        return {spec.key for spec in self.asset_specs}
+
+    @cached_property
+    def dag_ids(self) -> Set[str]:
+        return set(self.task_id_map.keys())
+
+    @cached_property
+    # dag_id -> set of task_ids
+    def task_id_map(self) -> Dict[str, Set[str]]:
+        task_id_map_data = {
+            dag_id: set(ta_map.keys()) for dag_id, ta_map in self.asset_key_map.items()
+        }
+        return defaultdict(set, task_id_map_data)
+
+    @cached_property
+    # dag_id -> set of asset_keys
+    def asset_keys_per_dag_id(self) -> Dict[str, Set[AssetKey]]:
+        asset_keys_per_dag_data = {
+            dag_id: {
+                asset_key for asset_keys in task_to_asset_map.values() for asset_key in asset_keys
+            }
+            for dag_id, task_to_asset_map in self.asset_key_map.items()
+        }
+        return defaultdict(set, asset_keys_per_dag_data)
+
+    @cached_property
+    # dag_id -> task_id -> set of asset_keys
+    def asset_key_map(self) -> Dict[str, Dict[str, Set[AssetKey]]]:
+        asset_key_map: Dict[str, Dict[str, Set[AssetKey]]] = defaultdict(lambda: defaultdict(set))
+        for spec in self.asset_specs:
+            if TASK_ID_METADATA_KEY in spec.metadata:
+                dag_id = spec.metadata[DAG_ID_METADATA_KEY]
+                task_id = spec.metadata[TASK_ID_METADATA_KEY]
+                asset_key_map[dag_id][task_id].add(spec.key)
+        return asset_key_map
+
+
+def build_task_spec_mapping_info(defs: Definitions) -> TaskSpecMappingInfo:
+    asset_specs = list(spec_iterator(defs.assets))
+    return TaskSpecMappingInfo(asset_specs=asset_specs)
+
+
 def compute_serialized_data(
     airflow_instance: AirflowInstance, defs: Definitions
 ) -> "SerializedAirflowDefinitionsData":
     migration_state = airflow_instance.get_migration_state()
+
+    task_spec_mapping_info = build_task_spec_mapping_info(defs)
+
     dag_infos = {dag.dag_id: dag for dag in airflow_instance.list_dags()}
 
     downstreams_asset_dependency_graph: Dict[AssetKey, Set[AssetKey]] = defaultdict(set)
```

Comment:
this code path definitely suffers from a "too many local variables" code smell

Reply:
if we did a similar functional cached property thing for the asset graph stuff, that would go away in a large way right?

Reply:
Seems like this entire loop can be replaced with that type of functional mapping I think. That can be a follow up though.

Reply:
Yah agree. Just doing it incrementally to accomplish the task at hand.